### ISSUE TITLE
PHP User Auth

### ISSUE THREAD

**Issue by Sledro** [kreait/firebase-php#159]
Hey, 

Is there a function to login a user via email and password such as the js method?

firebase.auth().signInWithEmailAndPassword(userEmail, userPass).catch(function(error)

Cheers!

**Comment by Sledro** [comment]
I am thinking I can use:

`$auth = $firebase->getAuth();

$auth->getUserByEmailAndPassword("test@test.com","password");`

But how can I then get the logged in users info?

**Comment by jeromegamez** [comment]
This should work:

```php
$user = $auth->getUserByEmailAndPassword("test@test.com","password");

$userInfo = $auth->getUserInfo($user->getUid());
```

**Comment by Sledro** [comment]
I owe you a 🍺. Awesome work. 

💡 Maybe you could put a section in the docs about authenticating users. Thank you!

**Comment by jeromegamez** [comment]
To be honest, this section has been in the documentation but I removed it on purpose :D, and here's why:

- When you instantiate a Firebase instance with the Admin SDK, you already have the permission to access *everything*, so you don't need to authenticate a user by signing them in with their clear text password, because you are already authenticated.

- When you want to access your Firebase project with the permission of a given user, you should use the `asUser(...)` method as described in https://firebase-php.readthedocs.io/en/stable/authentication.html#authenticate-with-limited-privileges - this will limit your access to the same access as the given user has when they login to Firebase.

- If you want to make sure that a user accessing your (non-firebase) application, you should let them login in your frontend application with one of the official frontend libraries from https://firebase.google.com/docs/guides/ - don't let them enter their password in your custom frontend, pass the password to your backend and then use the password to authenticate them via the PHP Admin SDK -> let them login via the official client library, get their ID Token and verify the ID token in your backend (see https://firebase-php.readthedocs.io/en/stable/authentication.html#verify-a-firebase-id-token).

Once the token is successfully verified, you can then use the `"sub"` claim from the token and don't have to work with passwords at all.

The only reason I kept the `getUserByEmailAndPassword()` method is to provide full backwards compatibility in the 3.x releases, but it will get deprecated as soon as I have implemented a proper `UserManager` and removed in 4.x 😇

**Comment by Sledro** [comment]
Ah OK fair enough, thanks for the detailed reply. I had a look and have been trying to get my head around what you are saying. I understand your first two points fully. As for the third... Is there any basic examples of this approach online?

So say I use the JS front end? and pass the password to the SDK and verify it with something like $user->verifyPass("password")? That must not be what you are saying as its basically a long winded way of doing what I am currently doing haha

I was trying to avoid the js front end for logging in.

**Comment by jeromegamez** [comment]
Perhaps I will keep the method, I haven't fully made up my mind yet, to be honest ¯\_(ツ)_/¯ 

The thing is: it's kind of a security risk. Instead of using the secure authentication mechanisms provided by Firebase itself, you are proxying it through your own app... Once the password lands on your own server, it's vulnerable to exposure - so I would at least not recommend it 😅

**Comment by Sledro** [comment]
I have gone about doing it your way. I have the js passing the tokenid to a file called verify-token.php.


```
$idTokenString = $_POST['variable']; // The token from the client

try {
    $verifiedIdToken = $firebase->getAuth()->verifyIdToken($idTokenString);
} catch (InvalidIdToken $e) {
    echo $e->getMessage();
}
```



 Its responding this this error:

`<br />
<b>Fatal error</b>:  Uncaught Kreait\Firebase\Exception\Auth\InvalidIdToken: This token has been issued in the future at 2018-01-25T01:36:31+00:00, is your system time correct? in /opt/lampp/htdocs/TotalFitness/Dashboard/vendor/kreait/firebase-php/src/Firebase/Auth/IdTokenVerifier.php:33
Stack trace:
#0 /opt/lampp/htdocs/TotalFitness/Dashboard/vendor/kreait/firebase-php/src/Firebase/Auth.php(175): Kreait\Firebase\Auth\IdTokenVerifier-&gt;verify('eyJhbGciOiJSUzI...')
#1 /opt/lampp/htdocs/TotalFitness/Dashboard/includes/verify-token.php(10): Kreait\Firebase\Auth-&gt;verifyIdToken('eyJhbGciOiJSUzI...')
#2 {main}
  thrown in <b>/opt/lampp/htdocs/TotalFitness/Dashboard/vendor/kreait/firebase-php/src/Firebase/Auth/IdTokenVerifier.php</b> on line <b>33</b><br />`


Any ideas why this would be?

**Comment by Sledro** [comment]
I found the same issue #140 

Did you get a chance to add?

```
try {
    $idToken = $auth->verifyIdToken($idTokenString);
} catch (IdTokenIssuedInTheFuture $e) {
    $idToken = $e->getUnverifiedToken(); // or $e->getInsecureIdToken();
}
```

**Comment by kaoz70** [comment]
Check my comment on issue #140 maybe that would help you guys

**Comment by jeromegamez** [comment]
@Sledro @kaoz70 Please consider moving to the 4.0 dev release for your tests. In the next Major Version, the SDK will make full use of the external token generator and verifier. With it, you can catch the distinct exceptions and deal with them as you wish, e.g. compare the timestamps and check for an allowed time difference.

Time comparisons during the token verification are done with Unix timestamps, which are UTC based and time zone independent.

**Comment by jeromegamez** [comment]
See #312 for more on this topic

**Comment by jeromegamez** [comment]
Please check #390 for the current state of implementation.

**Comment by jeromegamez** [comment]
Released with [4.41.0](https://github.com/kreait/firebase-php/releases/tag/4.41.0) 🚀
